sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala

```diff
@@ -117,14 +120,166 @@ private[hive] class HiveClientImpl(
     case hive.v3_1 => new Shim_v3_1()
   }
 
+  def getCurrentState(): SessionState = {
+    val ugi = UserGroupInformation.getCurrentUser
+    // TODO need to check if we should add more condition
+    if (ugi.getAuthenticationMethod() == UserGroupInformation.AuthenticationMethod.PROXY) {
+      getOrCreateSessionState(ugi.getShortUserName)
+    } else {
+      state
+    }
+  }
+
+  def getCurrentHiveClient(): Hive = {
+    val ugi = UserGroupInformation.getCurrentUser
+    // TODO need to check if we should add more condition
+    if (ugi.getAuthenticationMethod() == UserGroupInformation.AuthenticationMethod.PROXY) {
+      getOrCreateSessionHive(ugi.getShortUserName)
+    } else {
+      cachedClient
+    }
+  }
+
+  val sessionStateMap = new ConcurrentHashMap[String, SessionState]()
+  val sessionHiveMap = new ConcurrentHashMap[String, Hive]()
+
+  /**
+   * Get or create SessionState for client user. If there have a existed SessionState,
+   * return old SessionState, else create a new one with client user name.
+   *
+   * @param user client user name need to proxy
+   * @return
+   */
+  private def getOrCreateSessionState(user: String): SessionState = sessionStateMap.synchronized {
+    if (sessionStateMap.containsKey(user)) {
+      sessionStateMap.get(user)
+    } else {
+      val original = Thread.currentThread().getContextClassLoader
+      // Switch to the initClassLoader.
+      Thread.currentThread().setContextClassLoader(initClassLoader)
+      try {
+        Hive.closeCurrent()
+        val state = newState(user)
+        sessionStateMap.put(user, state)
+        state
+      } finally {
+        Thread.currentThread().setContextClassLoader(original)
+      }
+    }
+  }
+
+  /**
+   * Get or create a hive client for client user. If there have a existed hive client,
+   * return old client, else create a new one with proxy.
+   *
+   * @param user client user name need to proxy
+   * @return
+   */
+  private def getOrCreateSessionHive(user: String): Hive = sessionHiveMap.synchronized {
+    if (sessionHiveMap.containsKey(user)) {
+      sessionHiveMap.get(user)
+    } else {
+      val original = Thread.currentThread().getContextClassLoader
+      // Switch to the initClassLoader.
+      Thread.currentThread().setContextClassLoader(initClassLoader)
+      try {
+        Hive.closeCurrent()
+        val currentConf = conf
+        val ugi = UserGroupInformation.createProxyUser(user, UserGroupInformation.getLoginUser)
+        val hive = if (obtainDelegationTokenRequired(currentConf, ugi)) {
+          currentConf.set("hive.metastore.token.signature", IMPERSONATION_TOKEN_SIGNATURE)
+          val delegationTokenStr = getDelegationToken(user)
+          org.apache.hadoop.hive.shims.Utils.setTokenStr(ugi, delegationTokenStr,
```

Comment:
Use HiveMetastore DelegationToken to pass Kerberos check since there are two way for connect to a kerberized Hive Metastore:
1. Kerberied client
2. Delegation token (got by kerberized client for spec user ), in this way we need to config `hive.metastore.token.signature` and this way is more stable and Hiveserver2 use this way. just use UGI.doAs() is not  stable because of security context reuse.